gpbft/participant.go

```diff
@@ -151,6 +151,49 @@ func (p *Participant) ReceiveAlarm() (err error) {
 	return nil
 }
 
+// FinalityInfo is an auxiliary struct to pass relevant info from the finality certificate
+// NOTE: It is not clear to me what information is available in the finality certificate
+// and in what form. The information provided here may change, using it as a placeholder for now.
+type FinalityInfo struct {
+	Instance uint64
+	Power    *PowerTable
+	Beacon   []byte
+}
+
+func (p *Participant) ReceiveFinalityCertificate(f FinalityInfo) error {
+	// Get ready to start the instance defined in the finality certificate for the next instance
+	p.gpbft = nil
+	prevInstance := p.currentInstance
+	// Set the instance from the certificate as the current one.
+	p.currentInstance = f.Instance
+	// clean all messages queued for instances below the one in the finality certificate
+	// and old committees
+	for i := prevInstance + 1; i < p.currentInstance; i++ {
+		delete(p.mqueue.messages, i)
+		delete(p.committees, i)
+	}
+
+	// store the new committees provided by the finality certificate
+	// overwriting whatever was there already (if anything).
+	// NOTE: this may not be needed and explicitly pass the committee
+	// to this function
+	// as depending on how we handle certificates in the host, it may
+	// already store the new power tables to make them available for gpbft
+	// through the host without additional work.
+	if err := f.Power.Validate(); err != nil {
+		return fmt.Errorf("invalid power table: %w", err)
+	}
+	com := &committee{power: f.Power, beacon: f.Beacon}
+	p.committees[f.Instance] = com
```

Comment:
This will be fetched when the instance starts. Don't write it here (note it would also need mutex protection)

Reply:
Thank you for the feedback, I didn't know to what extent we could assume the committee to be already available in the host. Question solved, thanks!